### ISSUE TITLE
Futex variables are 32 bit (int) on Linux but ours are 64 bit

### ISSUE THREAD

**Issue by stefanberger** [oscarlab/graphene#1485]
The mutex_handle in https://github.com/oscarlab/graphene/blob/master/Pal/src/host/Linux/pal_host.h#L39 defines a 64 bit variable that is later on used as the futex variable and also used in cmpxchg here https://github.com/oscarlab/graphene/blob/master/Pal/src/host/Linux/db_mutex.c#L108 . Futex variables are 32 bit on Linux (int) though [http://man7.org/linux/man-pages/man2/futex.2.html] the cmpxchg works on 64 bit. I think we should fix this by changing the variable to be int32_t or int32_t * and having cmpxchg works on 32 bit variables or by introducing cmpxchg32.

Comments?

**Comment by dimakuv** [comment]
Yes, we need to fix this for sure by changing `mutex->locked` to 32 bit. We should also use the 32-bit version of cmpxchg.

However, I would also prefer to decide on which atomics we use in Graphene. Currently we have at least two options: GCC builtins (see  e.g. https://github.com/oscarlab/graphene/blob/31b752a3350ba250a941a153e15e5167e792df46/Pal/include/lib/spinlock.h#L162) and atomics stolen from Musl (https://github.com/oscarlab/graphene/blob/master/Pal/include/lib/x86_64/atomic.h).

I would definitely prefer GCC builtins. @yamahata @mkow @boryspoplawski Did we have a consensus on what we wanted to do with atomics?

If we use GCC builtins, then we can simply use `__atomic_compare_exchange_n()` on our 32-bit `mutex->locked`.

**Comment by boryspoplawski** [comment]
I wasn't part of any larger discussions about it, but I would prefer GCC builtins as well.

**Comment by mkow** [comment]
I thought we wanted to move to C11 atomics? Or is there a problem in using them without stdlib?

**Comment by boryspoplawski** [comment]
I think you cannot assume anything about how C11 atomics are represented i.e. `atomic int` could be a struct, which kind of makes it unusable in most places for us.

**Comment by mkow** [comment]
Ah, right, we're using those atomics on the buffer passed to `futex()`. Then GCC builtin sounds good, although they may bite us later when porting this to other compilers.

Ad. other places: isn't this only a problem when we use something both for futexes and operating on this manually? This is IMO a very limited use case in our codebase.

**Comment by dimakuv** [comment]
LLVM implements GCC builtins:
> The first set of library functions are named __atomic_*. This set has been “standardized” by GCC, and is described below. (see https://llvm.org/docs/Atomics.html).
